solr/core/src/java/org/apache/solr/query/DynamicComplementPrefixQuery.java

```diff
@@ -0,0 +1,353 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.query;
+
+
+import java.io.IOException;
+
+import org.apache.lucene.index.FilteredTermsEnum;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SortedDocValues;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.index.TermState;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.DocValuesFieldExistsQuery;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.MultiTermQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.AttributeSource;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.UnicodeUtil;
+
+/** A Query that matches documents containing terms with a specified prefix. A PrefixQuery
+ * is built by QueryParser for input like <code>app*</code>.
+ *
+ * <p>This query differs from {@link org.apache.lucene.search.PrefixQuery} in two ways:
+ * 1. It does _not_ build an automaton representing the prefix; rather, it builds
+ * {@link FilteredTermsEnum} instances that are optimized for common data patterns
+ * (specifically, it compares the _end_ of BytesRefs first).
+ * 2. In the event that the prefix matches a large proportion of terms in the index,
+ * an inverse query is built against terms that do _not_ match the prefix.
+ * */
+
+public class DynamicComplementPrefixQuery extends MultiTermQuery {
+
+  private final Term termPrefix;
+  private final BytesRef prefix;
+  private final BytesRef limit;
+  private final MultiTermQuery inverted;
+
+  /** Constructs a query for terms starting with <code>prefix</code>. */
+  public DynamicComplementPrefixQuery(Term prefix, boolean noInvert) {
+    // It's OK to pass unlimited maxDeterminizedStates: the automaton is born small and determinized:
```

Comment:
Should this comment be removed as this is not using automaton?

Reply:
thanks for catching this; fixed with e14b68661d1cdb34df51ce466e8b4673b52d32bd